File: core/numik.py
"""Numerical task-space inverse kinematics (MuJoCo damped least squares).

Replaces the closed-form Yi-2012 analytic IK with the approach the recent
literature converges on (Ryan 2025 / OmniDP 2026 / MIRROR 2026; Jiang 2025's
joint+Cartesian blend): solve joint angles on the *actual robot model* so the
arm reaches task targets. We drive BOTH the elbow and the wrist body to target
positions — matching the operator's upper-arm AND forearm directions at once.
This removes the 3-DOF analytic IK's bent-arm ambiguity (it could not satisfy
upper-arm direction and wrist position simultaneously).

Model-agnostic: works on any MuJoCo model (MJCF or a loaded URDF) given the
arm's joint names + shoulder/elbow/wrist body names. Jacobians come from
``mujoco.mj_jacBody`` so no kinematics are hard-coded — drop in a real URDF and
it adapts.
"""

from __future__ import annotations

from typing import Any

import numpy as np
from numpy.typing import NDArray


class ArmPositionIK:
    """Damped-least-squares IK driving one arm's elbow + wrist bodies to targets.

    Parameters
    ----------
    model:
        A ``mujoco.MjModel``.
    joint_names:
        Arm joints to actuate for the position task, base→tip. Typically the 4
        that move the wrist: shoulder pitch/roll/yaw + elbow. (Wrist twist/pitch
        are distal to the wrist body origin and don't change its position.)
    shoulder_body / elbow_body / wrist_body:
        Body names whose origins are the shoulder, elbow, and wrist (hand) points.
    damping:
        DLS Levenberg damping λ (m). Larger = more stable near singularities,
        slower convergence.
    max_iters, pos_tol, step_clip:
        Iteration budget, convergence tolerance (m), and per-iter joint step cap.
    """

    def __init__(
        self,
        model: Any,
        joint_names: list[str],
        shoulder_body: str,
        elbow_body: str,
        wrist_body: str,
        damping: float = 0.08,
        max_iters: int = 16,
        pos_tol: float = 2e-3,
        step_clip: float = 0.35,
        joint_limits: dict[str, tuple[float, float]] | None = None,
    ) -> None:
        import mujoco

        self._mj = mujoco
        self.model = model
        jids = [mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_JOINT, n) for n in joint_names]
        if any(j < 0 for j in jids):
            missing = [n for n, j in zip(joint_names, jids, strict=True) if j < 0]
            raise ValueError(f"joints not found in model: {missing}")
        self.joint_names = list(joint_names)
        self.qadr = np.array([model.jnt_qposadr[j] for j in jids], dtype=int)
        self.dof = np.array([model.jnt_dofadr[j] for j in jids], dtype=int)
        lo: list[float] = []
        hi: list[float] = []
        for j in jids:
            if bool(model.jnt_limited[j]):
                lo.append(float(model.jnt_range[j, 0]))
                hi.append(float(model.jnt_range[j, 1]))
            else:
                lo.append(-np.inf)
                hi.append(np.inf)
        self.lo = np.array(lo)
        self.hi = np.array(hi)
        # Dedicated joint-limit config (robot.joint_limits) overrides the model's
        # jnt_range so limits track the real robot independent of placeholder
        # model dims. Matched by full ("r_elbow_joint") or canonical ("r_elbow").
        if joint_limits:
            for i, n in enumerate(self.joint_names):
                canon = n[:-6] if n.endswith("_joint") else n
                lim = joint_limits.get(n, joint_limits.get(canon))
                if lim is not None:
                    self.lo[i] = float(lim[0])
                    self.hi[i] = float(lim[1])
        self.sb = mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_BODY, shoulder_body)
        self.eb = mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_BODY, elbow_body)
        self.wb = mujoco.mj_name2id(model, mujoco.mjtObj.mjOBJ_BODY, wrist_body)
        if min(self.sb, self.eb, self.wb) < 0:
            raise ValueError("shoulder/elbow/wrist body not found in model")
        self.damping = damping
        self.max_iters = max_iters
        self.pos_tol = pos_tol
        self.step_clip = step_clip
        # Distance (m) still separating the bodies from their targets after the
        # last :meth:`solve`. Callers gate on this: a solve that did not converge
        # has produced *a* pose, just not the requested one, and committing it is
        # worse than holding the arm.
        self.last_residual_m = 0.0
        # Only escape once the error is this large (m). Set by RobotModel from the
        # arm's own link lengths. Without it the escape also fires when DLS has
        # simply reached the best pose the arm can hold — 4 joints against 6
        # position constraints rarely leaves zero error — and then it nudges
        # joints that were doing fine. Measured on upperbody_sim.xml: a genuinely
        # stuck solve sits at 0.08-0.80 m while healthy convergence plateaus at
        # or below 0.03 m, so anything in between separates them.
        self.escape_error_m = 0.05
        # Deterministic singularity escape, alternating signs so the nudge is not
        # biased toward one side. See :meth:`solve`.
        self._escape_pattern = np.array(
            [1.0 if i % 2 == 0 else -1.0 for i in range(len(self.joint_names))]
        )

    # A joint whose task-Jacobian column is this short moves the targets by
    # essentially nothing, so DLS has no reason to turn it -- that is what a
    # rank-deficient configuration looks like from inside the loop.
    _WEAK_COLUMN = 1e-3      # metres of body motion per radian
    # Progress counts as stalled below the larger of these: an absolute floor and
    # a fraction of the remaining error.
    #
    # Deliberately NOT a test on ||dq||: at the shoulder singularity the elbow
    # keeps taking large useless steps (it folds to reach the wrist) while the
    # shoulder is frozen, so the iteration looks busy while going nowhere.
    #
    # The relative term matters just as much. DLS approaches a solution
    # asymptotically, so per-iteration progress falls below any fixed floor while
    # the error is still above ``pos_tol``. An absolute-only test fired during
    # ordinary convergence and nudged joints that were doing fine -- observed as
    # shoulder yaw ratcheting to its limit over successive frames of a lateral
    # arm raise, which then never came back.
    _MIN_IMPROVE_M = 1e-4
    _MIN_IMPROVE_RATIO = 0.01
    # ...and the stall has to PERSIST. A descent that passes close to a
    # singularity legitimately crawls for an iteration or two before the useful
    # joint builds up leverage; a true fixed point never recovers. Escaping on the
    # first slow iteration wrecked poses the plain solver reached to 0.02 deg.
    _STALL_ITERS = 4
    _ESCAPE_STEP = 0.08      # rad; nudge applied to leave the singular manifold
    _MAX_ESCAPES = 3

    def body_pos(self, data: Any, which: str) -> NDArray[np.float64]:
        bid = {"shoulder": self.sb, "elbow": self.eb, "wrist": self.wb}[which]
        return np.array(data.xpos[bid], dtype=np.float64)

    def solve(
        self,
        data: Any,
        elbow_target: NDArray[np.float64],
        wrist_target: NDArray[np.float64],
    ) -> dict[str, float]:
        """Iterate DLS from the current ``data.qpos`` (warm start) toward targets.

        Mutates ``data.qpos`` for the arm joints and returns {joint_name: angle}.
        Warm-starting from the previous frame's solution gives temporal smoothness
        and resolves redundancy implicitly (stays near the last pose).

        Sets :attr:`last_residual_m` so the caller can tell a converged solve from
        a pose the solver merely stopped at.

        **Singularity escape.** Warm-starting has a failure mode: some
        configurations are fixed points. On a shoulder whose pitch and yaw axes
        coincide while the upper arm lies along them, every column of the task
        Jacobian is zero for certain target directions, so ``dq`` comes out zero
        and the arm never moves again — not this iteration, not this frame, not
        any later frame, because the next warm start begins from the same
        configuration. Measured on ``models/upperbody_sim.xml``: from the rest
        pose, "point the arm forward" stayed 90 deg off for 60 consecutive frames
        with the shoulder joints at exactly 0.000, and raising ``max_iters`` from
        16 to 300 changed nothing. When the iteration stalls we therefore nudge
        the joints that are doing nothing, which leaves the singular manifold and
        lets DLS take over again.
        """
        mj = self._mj
        nv = self.model.nv
        jp = np.zeros((3, nv))
        I6 = np.eye(6)
        escapes = 0
        stalled_iters = 0
        prev_err = np.inf
        for _ in range(self.max_iters):
            mj.mj_forward(self.model, data)
            pe = data.xpos[self.eb]
            pw = data.xpos[self.wb]
            err = np.concatenate([elbow_target - pe, wrist_target - pw])
            err_norm = float(np.linalg.norm(err))
            if err_norm < self.pos_tol:
                break
            if err_norm > self.escape_error_m and (prev_err - err_norm) <= max(
                self._MIN_IMPROVE_M, self._MIN_IMPROVE_RATIO * err_norm
            ):
                stalled_iters += 1
            else:
                stalled_iters = 0
            prev_err = err_norm
            mj.mj_jacBody(self.model, data, jp, None, self.eb)
            je = jp[:, self.dof].copy()
            mj.mj_jacBody(self.model, data, jp, None, self.wb)
            jw = jp[:, self.dof].copy()
            jac = np.vstack([je, jw])  # 6 x n
            # dq = Jᵀ (J Jᵀ + λ²I)⁻¹ e
            dq = jac.T @ np.linalg.solve(jac @ jac.T + (self.damping ** 2) * I6, err)
            dq = np.clip(dq, -self.step_clip, self.step_clip)
            if stalled_iters >= self._STALL_ITERS:
                if escapes >= self._MAX_ESCAPES:
                    break  # genuinely unreachable; residual below tells the caller
                escapes += 1
                stalled_iters = 0
                # Nudge only the joints whose column is essentially zero — the
                # ones the rank deficiency is hiding. Nudging a joint that is
                # already contributing just fights the solver.
                weak = np.linalg.norm(jac, axis=0) < self._WEAK_COLUMN
                if not weak.any():
                    weak = np.ones(len(self.qadr), dtype=bool)
                dq = np.where(weak, self._escape_pattern * self._ESCAPE_STEP * escapes, 0.0)
            q = np.clip(data.qpos[self.qadr] + dq, self.lo, self.hi)
            data.qpos[self.qadr] = q
        # Final command clamp — guarantee the returned/applied angles are within
        # [lo, hi] even if the loop converged before an update this call.
        data.qpos[self.qadr] = np.clip(data.qpos[self.qadr], self.lo, self.hi)
        mj.mj_forward(self.model, data)
        self.last_residual_m = float(
            np.linalg.norm(
                np.concatenate(
                    [
                        elbow_target - data.xpos[self.eb],
                        wrist_target - data.xpos[self.wb],
                    ]
                )
            )
        )
        return {
            n: float(data.qpos[a])
            for n, a in zip(self.joint_names, self.qadr, strict=True)
        }

File: cpp/src/numik.cpp
#include "itu/numik.hpp"

#include <mujoco/mujoco.h>

#include <Eigen/Dense>
#include <algorithm>
#include <limits>
#include <stdexcept>

namespace itu {

namespace {
// Strip a trailing "_joint" so a config keyed by canonical name still matches.
std::string canonical(const std::string& n) {
  const std::string suffix = "_joint";
  if (n.size() > suffix.size() && n.compare(n.size() - suffix.size(), suffix.size(), suffix) == 0)
    return n.substr(0, n.size() - suffix.size());
  return n;
}
}  // namespace

ArmPositionIK::ArmPositionIK(const mjModel* model, std::vector<std::string> joint_names,
                             const std::string& shoulder_body, const std::string& elbow_body,
                             const std::string& wrist_body, double damping, int max_iters,
                             double pos_tol, double step_clip,
                             const std::map<std::string, std::pair<double, double>>* joint_limits)
    : model_(model),
      joint_names_(std::move(joint_names)),
      damping_(damping),
      max_iters_(max_iters),
      pos_tol_(pos_tol),
      step_clip_(step_clip) {
  const double inf = std::numeric_limits<double>::infinity();
  std::vector<std::string> missing;
  for (const auto& n : joint_names_) {
    const int jid = mj_name2id(model, mjOBJ_JOINT, n.c_str());
    if (jid < 0) {
      missing.push_back(n);
      continue;
    }
    qadr_.push_back(model->jnt_qposadr[jid]);
    dof_.push_back(model->jnt_dofadr[jid]);
    if (model->jnt_limited[jid]) {
      lo_.push_back(model->jnt_range[2 * jid + 0]);
      hi_.push_back(model->jnt_range[2 * jid + 1]);
    } else {
      lo_.push_back(-inf);
      hi_.push_back(inf);
    }
  }
  if (!missing.empty()) {
    std::string msg = "joints not found in model:";
    for (const auto& m : missing) msg += " " + m;
    throw std::invalid_argument(msg);
  }
  // Dedicated joint-limit config overrides the model's jnt_range.
  if (joint_limits != nullptr) {
    for (std::size_t i = 0; i < joint_names_.size(); ++i) {
      const std::string& full = joint_names_[i];
      auto it = joint_limits->find(full);
      if (it == joint_limits->end()) it = joint_limits->find(canonical(full));
      if (it != joint_limits->end()) {
        lo_[i] = it->second.first;
        hi_[i] = it->second.second;
      }
    }
  }
  sb_ = mj_name2id(model, mjOBJ_BODY, shoulder_body.c_str());
  eb_ = mj_name2id(model, mjOBJ_BODY, elbow_body.c_str());
  wb_ = mj_name2id(model, mjOBJ_BODY, wrist_body.c_str());
  if (std::min({sb_, eb_, wb_}) < 0)
    throw std::invalid_argument("shoulder/elbow/wrist body not found in model");
}

Eigen::Vector3d ArmPositionIK::body_pos(const mjData* data, const std::string& which) const {
  int bid = sb_;
  if (which == "elbow") bid = eb_;
  else if (which == "wrist") bid = wb_;
  return Eigen::Vector3d(data->xpos[3 * bid + 0], data->xpos[3 * bid + 1], data->xpos[3 * bid + 2]);
}

JointMap ArmPositionIK::solve(mjData* data, const Eigen::Vector3d& elbow_target,
                              const Eigen::Vector3d& wrist_target) const {
  const int nv = model_->nv;
  const int n = static_cast<int>(joint_names_.size());
  std::vector<mjtNum> jp(static_cast<std::size_t>(3) * nv);
  const Eigen::Matrix<double, 6, 6> I6 = Eigen::Matrix<double, 6, 6>::Identity();

  for (int iter = 0; iter < max_iters_; ++iter) {
    mj_forward(model_, data);
    const Eigen::Vector3d pe(data->xpos[3 * eb_ + 0], data->xpos[3 * eb_ + 1],
                             data->xpos[3 * eb_ + 2]);
    const Eigen::Vector3d pw(data->xpos[3 * wb_ + 0], data->xpos[3 * wb_ + 1],
                             data->xpos[3 * wb_ + 2]);
    Eigen::Matrix<double, 6, 1> err;
    err.head<3>() = elbow_target - pe;
    err.tail<3>() = wrist_target - pw;
    if (err.norm() < pos_tol_) break;

    Eigen::MatrixXd jac(6, n);
    mj_jacBody(model_, data, jp.data(), nullptr, eb_);
    for (int r = 0; r < 3; ++r)
      for (int k = 0; k < n; ++k) jac(r, k) = jp[static_cast<std::size_t>(r) * nv + dof_[k]];
    mj_jacBody(model_, data, jp.data(), nullptr, wb_);
    for (int r = 0; r < 3; ++r)
      for (int k = 0; k < n; ++k) jac(3 + r, k) = jp[static_cast<std::size_t>(r) * nv + dof_[k]];

    // dq = Jᵀ (J Jᵀ + λ²I)⁻¹ e
    const Eigen::Matrix<double, 6, 6> a = jac * jac.transpose() + (damping_ * damping_) * I6;
    const Eigen::Matrix<double, 6, 1> y = a.partialPivLu().solve(err);
    Eigen::VectorXd dq = jac.transpose() * y;
    for (int k = 0; k < n; ++k) {
      const double d = std::clamp(dq[k], -step_clip_, step_clip_);
      const double q = std::clamp(data->qpos[qadr_[k]] + d, lo_[k], hi_[k]);
      data->qpos[qadr_[k]] = q;
    }
  }
  // Final command clamp — guarantee returned/applied angles are within [lo, hi].
  for (int k = 0; k < n; ++k)
    data->qpos[qadr_[k]] = std::clamp(data->qpos[qadr_[k]], lo_[k], hi_[k]);
  mj_forward(model_, data);

  JointMap out;
  for (int k = 0; k < n; ++k) out[joint_names_[k]] = data->qpos[qadr_[k]];
  return out;
}

}  // namespace itu
